bot/src/authorisation.ts

```diff
@@ -0,0 +1,73 @@
+import getSigner from 'auctions-core/src/signer';
+import {
+    authorizeCollateral,
+    authorizeWallet,
+    getCollateralAuthorizationStatus,
+    getWalletAuthorizationStatus,
+} from 'auctions-core/src/authorizations';
+import { ETHEREUM_NETWORK, KEEPER_PREAUTHORIZE, WHITELISTED_COLLATERALS } from './variables';
+import { parseCollateralWhitelist } from './whitelist';
+
+export async function authorizeKeeperWallet(authorizeCallback?: () => void) {
+    const signer = await getSigner(ETHEREUM_NETWORK);
+    const walletAddress = await signer.getAddress();
+    const isWalletAuth = await getWalletAuthorizationStatus(ETHEREUM_NETWORK, walletAddress);
+
+    if (!isWalletAuth) {
+        console.info(`keeper: wallet "${walletAddress}" has not been authorized yet. Attempting authorization now...`);
+        const transactionHash = await authorizeWallet(ETHEREUM_NETWORK, walletAddress, false);
+        console.info(`keeper: wallet "${walletAddress}" successfully authorized via "${transactionHash}" transaction`);
+        if (authorizeCallback) {
+            authorizeCallback();
+        }
+    }
+}
+
+export async function checkAndAuthorizeKeeperCollateral(
+    walletAddress: string,
+    collateralType: string,
+    authorizeCallback?: () => void
+) {
+    // get collateral authorization status
+    const isCollateralAuth = await getCollateralAuthorizationStatus(ETHEREUM_NETWORK, collateralType, walletAddress);
+
+    // try to authorize the collateral then return
+    if (!isCollateralAuth) {
+        console.info(
+            `keeper: collateral "${collateralType}" has not been authorized on wallet "${walletAddress}" yet. Attempting authorization now...`
+        );
+        const collateralTransactionHash = await authorizeCollateral(
+            ETHEREUM_NETWORK,
+            walletAddress,
+            collateralType,
+            false
+        );
+        console.info(
+            `keeper: collateral "${collateralType}" successfully authorized on wallet "${walletAddress}" via "${collateralTransactionHash}" transaction`
+        );
+        if (authorizeCallback) {
+            authorizeCallback();
+        }
+    }
+}
+
+export async function setupCollateralAuthorizations() {
+    if (!KEEPER_PREAUTHORIZE || !WHITELISTED_COLLATERALS) {
+        return;
+    }
+
+    // check if the wallet is authorized
+    await authorizeKeeperWallet();
+
+    const collaterals = parseCollateralWhitelist(WHITELISTED_COLLATERALS);
+    console.info(
+        `keeper: "KEEPER_PREAUTHORIZE" is true attempting to authorize collaterals: '${collaterals.toString()}'`
+    );
```

Comment:
In case there is no `WHITELISTED_COLLATERALS` this will fail right? But the issue expect preauth to work even if no whitelisted collaterals specified

Reply:
Yes, it returns if either `WHITELISTED_COLLATERALS` is not set or `KEEPER_PREAUTHORIZE` is false. 

As this part of the function is also ran after validating the whitelist, we can be sure, that the collaterals entered exist and are correct.

Reply:
I meant that pre-authorisation should happen on **all collaterals** if no `WHITELISTED_COLLATERALS` are specified. 

I also suggest that instead of parsing `WHITELISTED_COLLATERALS` every time you need them, you should have one place where you define something like `getCurrentCollaterals` (that does not require passing env variable) and receive list of collaterals that are either all or whitelisted.